### PR TITLE
OSSM-3912 Unify install bookinfo and install/upgrade smcp test

### PR DESCRIPTION
**Pull request by fjglira**
Jira: https://issues.redhat.com/browse/OSSM-3912 
I unified test smcp install and upgrade and installation bookinfo test case in only one test case with sub test cases that verify that after upgrade bookinfo application continues working properly.

### CODE REVIEW

File: pkg/tests/ossm/smoke_test.go [hunk 1]
```diff
@@ -19,50 +19,195 @@ import (
...
+			t.Logf("This test checks whether SMCP %s deletion delete all the resources", env.GetSMCPVersion())
+
+			t.LogStep("Delete SMCP and verify if this deletes all resources")
+			assertUninstallDeletesAllResources(t, env.GetSMCPVersion())
+		})
 	})
 }
 
+func validateStartUpProxyTime(t TestHelper, ratingYaml string) {
+	proxyStartTime, proxyLastTransitionTime := ExtractProxyTimes(t, ratingYaml)
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:151]:
```suggestion
	proxyStartTime, proxyReadyTime := ExtractProxyTimes(t, ratingYaml)
```

File: pkg/tests/ossm/smoke_test.go [hunk 2]
```diff
@@ -19,50 +19,195 @@ import (
...
+	startupTime := proxyLastTransitionTime.Sub(proxyStartTime)
+	t.Logf("Proxy startup time: %s", startupTime.String())
+	if startupTime > 10*time.Second {
+		t.Fatalf("Proxy startup time is too long: %s", startupTime.String())
+	}
+}
+
+func ExtractProxyTimes(t TestHelper, yamlString string) (time.Time, time.Time) {
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:159]: This would be much shorter and simpler if you changed `oc.GetYaml()` to accept `jsonPath string` instead of `checks ...common.CheckFunc` to directly extract the two timestamps. As far as I can see, we never actually pass any check functions, so we can get rid of this parameter.

File: pkg/tests/ossm/smoke_test.go [hunk 3]
```diff
@@ -19,50 +19,195 @@ import (
...
+						"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
+			})
+
+			t.LogStep("verify proxy startup time. Expected to be less than 10 seconds")
+			t.Log("Jira related: https://issues.redhat.com/browse/OSSM-3586")
+			t.Log("From proxy yaml, verify the time between status.containerStatuses.state.running.startedAt and status.conditions[type=Ready].lastTransitionTime")
+			ratingPod := pod.MatchingSelector("app=ratings", ns)(t, oc.DefaultOC)
+			ratingYaml := oc.GetYaml(t, ns, "pod", ratingPod.Name)
+			t.Log("Validate from YAML, the proxy startup time is less than 10 seconds for ratings pod")
+			validateStartUpProxyTime(t, ratingYaml)
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:88]: Should this be a separate subtest? It has its own description, related Jira, and it's a completely separate check than what the rest of this subtest checks.

File: pkg/tests/ossm/smoke_test.go [hunk 4]
```diff
@@ -19,50 +19,195 @@ import (
...
-				"HTTP header 'x-envoy-decorator-operation' is present in the response",
-				"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
+		t.NewSubTest(fmt.Sprintf("upgrade %s to %s", fromVersion, toVersion)).Run(func(t TestHelper) {
+			t.Logf("This test checks whether SMCP becomes ready after it's upgraded from %s to %s and bookinfo is still working after the upgrade", fromVersion, toVersion)
+
+			t.Cleanup(func() {
+				app.Uninstall(t, app.Bookinfo(ns), app.SleepNoSidecar(ns))
+			})
+
+			t.LogStepf("Upgrade SMCP from %s to %s", fromVersion, toVersion)
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:98]: I wonder: Given that we must support running control plane version x.y with proxy version x.(y-1), shouldn't we (also) run all the entire test suite with this scenario (using the two different versions). This is the only way we can be sure that the behavior is correct. It's critical that everything works correctly even if there's a version difference between the control plane and the data plane. Or do we already test this?

File: pkg/tests/ossm/smoke_test.go [hunk 5]
```diff
@@ -19,50 +19,195 @@ import (
...
+						"ProductPage didn't return 200 OK"),
+					assert.OutputContains(
+						"server: istio-envoy",
+						"HTTP header 'server: istio-envoy' is present in the response",
+						"HTTP header 'server: istio-envoy' is missing from the response"),
+					assert.OutputContains(
+						"x-envoy-decorator-operation",
+						"HTTP header 'x-envoy-decorator-operation' is present in the response",
+						"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
+			})
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:118]: This code is duplicated three times in this test. Should be in its own function (e.g. `assertTrafficFlowsThroughProxy()`).

File: pkg/tests/ossm/smoke_test.go [hunk 6]
```diff
@@ -19,50 +19,195 @@ import (
...
+
+			t.LogStep("Delete Bookinfo pods to validate proxy is still working after recreation and upgrade")
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:120]: Do we also need to check that the dataplane works correctly even if the control plane is offline (e.g. by deleting the control plane and checking that the traffic still flows through the proxy?)

Maybe we can do this in the last subtest, where we delete the SMCP.

**Review by luksa — COMMENTED** [review]
LGTM apart from some minor nits and questions about other tests.

File: pkg/tests/ossm/smoke_test.go (hunk 3 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:88]: Yes, I can do this and make a separated test

File: pkg/tests/ossm/smoke_test.go (hunk 4 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:98]: We are doing that test because we made the upgrade, verify that the traffic is working and them restart the pods to get the new proxy and verify agains the traffic. So this scenario is covered:
```
t.LogStep("Check if bookinfo productpage is running through the Proxy after the upgrade")
			retry.UntilSuccess(t, func(t test.TestHelper) {
				oc.Exec(t,
					pod.MatchingSelector("app=sleep", ns), "sleep",
					"curl -sI http://productpage:9080",
					assert.OutputContains(
						"HTTP/1.1 200 OK",
						"ProductPage returns 200 OK",
						"ProductPage didn't return 200 OK"),
					assert.OutputContains(
						"server: istio-envoy",
						"HTTP header 'server: istio-envoy' is present in the response",
						"HTTP header 'server: istio-envoy' is missing from the response"),
					assert.OutputContains(
						"x-envoy-decorator-operation",
						"HTTP header 'x-envoy-decorator-operation' is present in the response",
						"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
			})

			t.LogStep("Delete Bookinfo pods to validate proxy is still working after recreation and upgrade")
			oc.RestartAllPodsAndWaitReady(t, ns)
			retry.UntilSuccess(t, func(t test.TestHelper) {
				oc.Exec(t,
					pod.MatchingSelector("app=sleep", ns), "sleep",
					"curl -sI http://productpage:9080",
					assert.OutputContains(
						"HTTP/1.1 200 OK",
						"ProductPage returns 200 OK",
						"ProductPage didn't return 200 OK"),
					assert.OutputContains(
						"server: istio-envoy",
						"HTTP header 'server: istio-envoy' is present in the response",
						"HTTP header 'server: istio-envoy' is missing from the response"),
					assert.OutputContains(
						"x-envoy-decorator-operation",
						"HTTP header 'x-envoy-decorator-operation' is present in the response",
						"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
			})
```

File: pkg/tests/ossm/smoke_test.go (hunk 6 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:120]: Makes sense, will aded

File: pkg/tests/ossm/smoke_test.go (hunk 5 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:118]: Done

File: pkg/tests/ossm/smoke_test.go (hunk 2 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:159]: I need to open a new Jira for this, because cluster wide test use, so I need to make change in more than one test at the same time. https://issues.redhat.com/browse/OSSM-4032

Comment (luksa) [reply pkg/tests/ossm/smoke_test.go:157]: Actually, if you pass in jsonpath, you get back JSON, not YAML, so you could simply create a new function: `oc.GetJSON(..., jsonPath string)`. So, you don't have to change any existing code.

File: pkg/tests/ossm/smoke_test.go (hunk 3 above)
Comment (luksa) [reply pkg/tests/ossm/smoke_test.go:74]: Btw, when someone comments on a PR, you should only reply instead of also clicking on "Resolve conversation", so that the person can easily see your reply without having to "Show resolved". The original comment author should be the one to resolve a conversation.

File: pkg/tests/ossm/smoke_test.go (hunk 4 above)
Comment (luksa) [reply pkg/tests/ossm/smoke_test.go:84]: Yes, but this only checks the very basic use-case (whether the traffic flows through the proxy). It doesn't check everything else that can be configured (e.g. traffic splitting, circuit breaking, mTLS, etc.). How do we know those features work properly when the control plane version is not the same as the data plane version?

Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:84]: But to make this happens we will increase to the double the execution time, yes will be more coverage and verification. But maybe this can be a feature to run in night builds where time does not care at the same level than in daily builds. wdyt?

File: pkg/tests/ossm/smoke_test.go [hunk 7]
```diff
@@ -19,50 +19,193 @@ import (
...
+						"ProductPage returns 200 OK",
+						"ProductPage didn't return 200 OK"),
+					assert.OutputContains(
+						"server: istio-envoy",
+						"HTTP header 'server: istio-envoy' is present in the response",
+						"HTTP header 'server: istio-envoy' is missing from the response"),
+					assert.OutputContains(
+						"x-envoy-decorator-operation",
+						"HTTP header 'x-envoy-decorator-operation' is present in the response",
+						"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:109]: Isn't this also the same code that's in `assertTrafficFlowsThroughProxy`?

File: pkg/tests/ossm/smoke_test.go [hunk 8]
```diff
@@ -19,50 +19,193 @@ import (
...
+			assertSMCPDeploysAndIsReady(t, fromVersion)
+
+			t.LogStep("Install bookinfo pods with sidecar and sleep pod without sidecar")
+			app.InstallAndWaitReady(t, app.Bookinfo(ns), app.SleepNoSidecar(ns))
+
+			t.LogStep("Check whether sidecar is injected in all bookinfo pods")
+			assertSidecarInjectedInAllBookinfoPods(t, ns)
+
+			t.LogStep("Check if bookinfo productpage is running through the Proxy")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:64]: You can probably move this retry into `assertTrafficFlowsThroughProxy`.

File: pkg/tests/ossm/smoke_test.go (hunk 7 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:109]: Yes, is the same. I'll move it to another place to reuse them

File: pkg/tests/ossm/smoke_test.go (hunk 8 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:64]: Done

File: pkg/tests/ossm/smoke_test.go [hunk 9]
```diff
@@ -19,50 +19,193 @@ import (
...
-			assert.OutputContains(
-				"x-envoy-decorator-operation",
-				"HTTP header 'x-envoy-decorator-operation' is present in the response",
-				"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
+		t.NewSubTest(fmt.Sprintf("upgrade %s to %s", fromVersion, toVersion)).Run(func(t TestHelper) {
+			t.Logf("This test checks whether SMCP becomes ready after it's upgraded from %s to %s and bookinfo is still working after the upgrade", fromVersion, toVersion)
+
+			t.Cleanup(func() {
+				app.Uninstall(t, app.Bookinfo(ns), app.SleepNoSidecar(ns))
+			})
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:80]: This should probably be moved to the test (not subtest). Also, these apps should then be installed in the test (not subtest) and pods restarted after the SMCP is ready.

Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:80]: Done

File: pkg/tests/ossm/smoke_test.go (hunk 4 above)
Comment (luksa) [reply pkg/tests/ossm/smoke_test.go:92]: Yes, this should be a separate test run, just like the upgrade, interop, and other test types. Ideally, we would add two additional test runs: with control plane version being 1 more than the data plane version, and with the control plane version being 1 less than the data plane version (to test downgrades).

**Comment by fjglira** [comment]
> I just saw `delete_smcp_v2.4` subtest fail, because the `route.route.openshift.io/bookinfo-bookinfo-gateway-525eca1d5089dbdc` wasn't deleted.

Yes, I need to make some fix.

File: pkg/tests/ossm/smoke_test.go [hunk 10]
```diff
@@ -19,32 +19,101 @@ import (
...
+		&version.SMCP_2_2,
+		&version.SMCP_2_3,
+		&version.SMCP_2_4,
+	}
+)
+
+type PodTime struct {
+	PodName            string
+	StartedAt          time.Time
+	LastTransitionTime time.Time
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:49]:
```suggestion
	ReadyAt time.Time
```

File: pkg/tests/ossm/smoke_test.go [hunk 11]
```diff
@@ -19,32 +19,101 @@ import (
...
+			})
+
+			t.LogStep("Check if bookinfo productpage is running through the Proxy")
+			assertTrafficFlowsThroughProxy(t, ns)
+
+			t.LogStep("verify proxy startup time. Expected to be less than 10 seconds")
+			t.Log("Jira related: https://issues.redhat.com/browse/OSSM-3586")
+			t.Log("From proxy json , verify the time between status.containerStatuses.state.running.startedAt and status.conditions[type=Ready].lastTransitionTime")
+			t.Log("The proxy startup time should be less than 10 seconds for ratings pod")
+			validateStartUpProxyTime(t, ns)
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:88]:
```suggestion
			assertProxiesReadyInLessThan10Seconds(t, ns)
```

(or pass 10s as a parameter and rename the function to `assertProxiesReadyIn()`).

File: pkg/tests/ossm/smoke_test.go [hunk 12]
```diff
@@ -63,6 +132,56 @@ func TestBookinfoInjection(t *testing.T) {
...
+	for _, podTime := range podTimes {
+		startupTime := podTime.LastTransitionTime.Sub(podTime.StartedAt)
+		t.Logf("Proxy startup time: %s", startupTime.String())
+		if startupTime > 10*time.Second {
+			t.Fatalf("Proxy startup time is too long: %s", startupTime.String())
+		}
+	}
+}
+
+func ExtractProxyTimes(t TestHelper, ns string) []PodTime {
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:146]: What's the benefit of extracting times into a struct and then checking. You could simply get the times and check them directly (I mean, merge validateStartupProxyTime and ExtractProxyTimes into a single function).

File: pkg/tests/ossm/smoke_test.go [hunk 13]
```diff
@@ -63,6 +132,56 @@ func TestBookinfoInjection(t *testing.T) {
...
+		}
+
+		podTime := PodTime{
+			PodName:            podName,
+			StartedAt:          startedAt,
+			LastTransitionTime: lastTransitionTime,
+		}
+		podTimes = append(podTimes, podTime)
+
+		t.Logf("Pod %s startedAt: %s, lastTransitionTime: %s", podName, startedAt.String(), lastTransitionTime.String())
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:179]: Do we need this log statement?

File: pkg/tests/ossm/smoke_test.go [hunk 14]
```diff
@@ -63,6 +132,56 @@ func TestBookinfoInjection(t *testing.T) {
...
+	scanner := bufio.NewScanner(strings.NewReader(proxyTimeList))
+	for scanner.Scan() {
+		line := scanner.Text()
+		fields := strings.Fields(line)
+		podName := fields[0]
+
+		// Skip if we have less than 3 fields, because we expect 3 fields per line
+		if len(fields) < 3 {
+			continue
+		}
```
Comment (luksa) [inline pkg/tests/ossm/smoke_test.go:161]: Does this ever happen?

**Review by luksa — COMMENTED** [review]
The test does what it should, but could be simplified a lot (there's a lot of code for simply checking the proxy startup time).

File: pkg/util/oc/oc_struct.go [hunk 15]
```diff
@@ -421,42 +421,43 @@ func (o OC) GetYaml(t test.TestHelper, ns, kind, name string, checks ...common.C
...
-	type ProxyResource struct {
-		Status *Proxy `json:"status"`
-	}
-
-	proxyResource := ProxyResource{}
+	proxy := o.GetJson(t, "", "proxy", "cluster", "{.status}")
 
-	proxyYaml := o.GetJson(t, "", "proxy", "cluster")
-	err := json.Unmarshal([]byte(proxyYaml), &proxyResource)
+	proxyStruct := &Proxy{}
```
Comment (luksa) [inline pkg/util/oc/oc_struct.go:439]: Why did you rename this from `proxy` to `proxyStruct`? It just makes the code harder to read.

File: pkg/util/oc/oc_struct.go [hunk 16]
```diff
@@ -421,42 +421,43 @@ func (o OC) GetYaml(t test.TestHelper, ns, kind, name string, checks ...common.C
...
+	var data map[string]interface{}
+	err := json.Unmarshal([]byte(proxy), &data)
 	if err != nil {
-		t.Fatalf("Could not parse Proxy resource: %v", err)
+		fmt.Printf("Failed to parse JSON: %s\n", err)
```
Comment (luksa) [inline pkg/util/oc/oc_struct.go:443]: Why doesn't this fail the test?

File: pkg/tests/ossm/smoke_test.go (hunk 13 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:179]: I put it here for testing reasons only

File: pkg/tests/ossm/smoke_test.go (hunk 14 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:161]: Yes, because in bookinfo namespace there is a sleep pod without proxy so the length of that line is 2

File: pkg/util/oc/oc_struct.go [hunk 17]
```diff
@@ -421,42 +421,43 @@ func (o OC) GetYaml(t test.TestHelper, ns, kind, name string, checks ...common.C
...
-		val = shell.Execute(t, fmt.Sprintf("oc %s get %s/%s -ojson", nsFlag(ns), kind, name), checks...)
+		jsonString = shell.Execute(t, fmt.Sprintf("oc %s get %s %s -o jsonpath='%s'", nsFlag(ns), kind, name, jsonPath))
 	})
-	return val
+
+	return jsonString
 }
 
 // GetProxy returns the Proxy object from the cluster
 func (o OC) GetProxy(t test.TestHelper) *Proxy {
```
Comment (luksa) [inline pkg/util/oc/oc_struct.go:436]: This function is now longer than it was initially, when it should end up being shorter because of the new jsonPath support.

It never makes sense to make the code longer if you're not actually adding any new code.

File: pkg/tests/ossm/smoke_test.go (hunk 13 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:179]: Deleted

File: pkg/tests/ossm/smoke_test.go (hunk 12 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go]: Done, create only one func and make it cleaner

File: pkg/util/oc/oc_struct.go (hunk 15 above)
Comment (fjglira) [reply pkg/util/oc/oc_struct.go:439]: Done

File: pkg/util/oc/oc_struct.go (hunk 16 above)
Comment (fjglira) [reply pkg/util/oc/oc_struct.go:443]: Done

File: pkg/util/oc/oc_struct.go (hunk 17 above)
Comment (fjglira) [reply pkg/util/oc/oc_struct.go:435]: Done

**Comment by fjglira** [comment]
Done all the changes

File: pkg/util/oc/oc_struct.go [hunk 18]
```diff
@@ -421,41 +421,41 @@ func (o OC) GetYaml(t test.TestHelper, ns, kind, name string, checks ...common.C
...
 	if err != nil {
-		t.Fatalf("Could not parse Proxy resource: %v", err)
+		t.Fatalf("Failed to parse JSON: %s\n", err)
 	}
-
-	proxy := proxyResource.Status
-	if proxy.HTTPProxy != "" {
-		t.Logf("HTTP_PROXY: %q", proxy.HTTPProxy)
+	if data["httpProxy"] == nil {
+		proxy.HTTPProxy = ""
```
Comment (luksa) [inline pkg/util/oc/oc_struct.go:445]: Why do you need to set this. It's already "".

File: pkg/util/oc/oc_struct.go [hunk 19]
```diff
@@ -421,41 +421,41 @@ func (o OC) GetYaml(t test.TestHelper, ns, kind, name string, checks ...common.C
...
 
-func (o OC) GetJson(t test.TestHelper, ns, kind, name string, checks ...common.CheckFunc) string {
+func (o OC) GetJson(t test.TestHelper, ns, kind, name, jsonPath string) string {
 	t.T().Helper()
-	var val string
+	var jsonString string
 	o.withKubeconfig(t, func() {
 		t.T().Helper()
-		val = shell.Execute(t, fmt.Sprintf("oc %s get %s/%s -ojson", nsFlag(ns), kind, name), checks...)
+		jsonString = shell.Execute(t, fmt.Sprintf("oc %s get %s %s -o jsonpath='%s'", nsFlag(ns), kind, name, jsonPath))
```
Comment (luksa) [inline pkg/util/oc/oc_struct.go:429]: What happens if jsonPath is ""? Does it return the whole JSON? If not, you should check for "" and use `-o json` in that case.

Comment (fjglira) [reply pkg/util/oc/oc_struct.go:429]: Ok, this has sense. I'll make the change

Comment (fjglira) [reply pkg/util/oc/oc_struct.go]: Done

File: pkg/tests/ossm/smoke_test.go [hunk 20]
```diff
@@ -63,6 +126,38 @@ func TestBookinfoInjection(t *testing.T) {
 	})
 }
 
+func assertProxiesReadyInLessThan10Seconds(t TestHelper, ns string) {
+	t.Log("Extracting proxy startup time and last transition time for all the pods in the namespace")
+	proxyTimeList := oc.GetJson(t, ns, "pods", "", `{range .items[*]}{.metadata.name}{"\t"}{.status.containerStatuses[?(@.name=="istio-proxy")].state.running.startedAt}{"\t"}{.status.conditions[?(@.type=="Ready")].lastTransitionTime}{"\n"}{end}`)
```
Comment (jewertow) [inline pkg/tests/ossm/smoke_test.go:131]: This command is tricky and the parser below is even much trickier. I had to execute this command and see the output to understand what it does. What do you think about something like this?
```suggestion
	startedAt := oc.GetJson(t, ns, "pods", "", `{range .items[*]}{.status.containerStatuses[?(@.name=="istio-proxy")].state.running.startedAt}{end}`)
	readyAt := oc.GetJson(t, ns, "pods", "", `{range .items[*]}{.status.conditions[?(@.type=="Ready")].lastTransitionTime}{end}`)
	startedAt = time.Parse(...)
	readyAt = time.Parse(...)
	startupTime := readyAt.Sub(startedAt)
	if startupTime > 10*time.Second {
        	t.Fatalf("Proxy startup time is too long: %s", startupTime.String())
	}
```

Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go]: Sure, better in that way. More easy to read

**Comment by fjglira** [comment]
@jewertow I updated the test case with you suggestion and it's working the test case:
```
--- PASS: TestSmoke (162.89s)
    --- PASS: TestSmoke/install_bookinfo_with_smcp_v2.3 (36.27s)
    --- PASS: TestSmoke/upgrade_v2.3_to_v2.4 (47.50s)
    --- PASS: TestSmoke/delete_smcp_v2.4 (38.91s)
PASS
ok  	github.com/maistra/maistra-test-tool/pkg/tests/ossm	164.810s
```

File: pkg/tests/ossm/smoke_test.go [hunk 21]
```diff
@@ -63,6 +126,36 @@ func TestBookinfoInjection(t *testing.T) {
 	})
 }
 
+func assertProxiesReadyInLessThan10Seconds(t TestHelper, ns string) {
+	t.Log("Extracting proxy startup time and last transition time for all the pods in the namespace")
+	startedAt := oc.GetJson(t, ns, "pods", "", `{range .items[*]}{.status.containerStatuses[?(@.name=="istio-proxy")].state.running.startedAt}{"\n"}{end}`)
+	readyAt := oc.GetJson(t, ns, "pods", "", `{range .items[*]}{.status.conditions[?(@.type=="Ready")].lastTransitionTime}{"\n"}{end}`)
```
Comment (jewertow) [inline pkg/tests/ossm/smoke_test.go:132]: Now I wonder whether the pods are always returned in the same order...? Maybe it would be better to fetch pods, iterate over them and fetch `startedAt` and `readyAt` for each pod individually. Then we would be 100% sure that a give pair of `startedAt` and `readyAt` refers to the same pod. What do you think?

File: pkg/tests/ossm/smoke_test.go [hunk 22]
```diff
@@ -63,6 +126,36 @@ func TestBookinfoInjection(t *testing.T) {
...
+
+	startedAtList := strings.Split(startedAt, "\n")
+	readyAtList := strings.Split(readyAt, "\n")
+
+	for i := 0; i < len(startedAtList) && i < len(readyAtList); i++ {
+		startedAtLine := startedAtList[i]
+		readyAtLine := readyAtList[i]
+		if startedAtLine == "" || readyAtLine == "" {
```
Comment (jewertow) [inline pkg/tests/ossm/smoke_test.go:140]: If `startedAt` or `readyAt` are empty, doesn't it mean that something is wrong with pods and we should retry fetching these values?

File: pkg/tests/ossm/smoke_test.go [hunk 23]
```diff
@@ -63,6 +126,36 @@ func TestBookinfoInjection(t *testing.T) {
...
+		podStartedAt, err := time.Parse(time.RFC3339, startedAtLine)
+		if err != nil {
+			t.Fatalf("Error parsing time for pod %d: %s", i, err)
+		}
+		podReadyAt, err := time.Parse(time.RFC3339, readyAtLine)
+		if err != nil {
+			t.Fatalf("Error parsing time for pod %d: %s", i, err)
+		}
+		startupTime := podReadyAt.Sub(podStartedAt)
+		if startupTime > 10*time.Second {
```
Comment (jewertow) [inline pkg/tests/ossm/smoke_test.go:153]: Actually, 10 seconds is fairly long for a test cluster. istio-proxy should start within ~1-2 seconds, so we might check whether proxies become ready in less than 3 seconds. What do you think?

File: pkg/tests/ossm/smoke_test.go (hunk 21 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:132]: Yes, I think will be better. I'll do the change

File: pkg/tests/ossm/smoke_test.go (hunk 22 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:140]: making the request of all the pods we will get rid of this, because the only pod without proxy is the sleep pod. So we can skip that pod verification

File: pkg/util/oc/oc_struct.go [hunk 24]
```diff
@@ -421,41 +421,45 @@ func (o OC) GetYaml(t test.TestHelper, ns, kind, name string, checks ...common.C
...
 	t.T().Helper()
-	var val string
+	var jsonString string
 	o.withKubeconfig(t, func() {
 		t.T().Helper()
-		val = shell.Execute(t, fmt.Sprintf("oc %s get %s/%s -ojson", nsFlag(ns), kind, name), checks...)
+		if jsonPath == "" {
+			jsonString = shell.Execute(t, fmt.Sprintf("oc %s get %s %s -o json", nsFlag(ns), kind, name))
+		} else {
+			jsonString = shell.Execute(t, fmt.Sprintf("oc %s get %s %s -o jsonpath='%s'", nsFlag(ns), kind, name, jsonPath))
```
Comment (luksa) [inline pkg/util/oc/oc_struct.go:432]: What if `jsonPath` contains single quotes?

File: pkg/util/oc/oc_struct.go (hunk 18 above)
Comment (luksa) [reply pkg/util/oc/oc_struct.go:449]: This code is still here.

File: pkg/tests/ossm/smoke_test.go [hunk 25]
```diff
@@ -63,6 +126,36 @@ func TestBookinfoInjection(t *testing.T) {
...
+			podStartedAt, err := time.Parse(time.RFC3339, startedAt)
+			if err != nil {
+				t.Fatalf("Error parsing time for pod %d", podName)
+			}
+			podReadyAt, err := time.Parse(time.RFC3339, readyAt)
+			if err != nil {
+				t.Fatalf("Error parsing time for pod %d", podName)
+			}
+			startupTime := podReadyAt.Sub(podStartedAt)
+			if startupTime > 3*time.Second {
```
[the remark this answers is not in the document]
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:150]: Right now in all the test that I did take less than 1 second. because of that I put 3 seconds

File: pkg/tests/ossm/smoke_test.go [hunk 26]
```diff
@@ -63,6 +126,36 @@ func TestBookinfoInjection(t *testing.T) {
 	})
 }
 
+func assertProxiesReadyInLessThan3Seconds(t TestHelper, ns string) {
+	t.Log("Extracting proxy startup time and last transition time for all the pods in the namespace")
+	podsList := oc.GetJson(t, ns, "pods", "", `{range .items[*]}{.metadata.name}{"\n"}{end}`)
+
+	for _, podName := range strings.Split(podsList, "\n") {
```
[the remark this answers is not in the document]
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:133]: Because of two things:

1.  In that namespace there is a pod without proxy (sleep pod), skip a pod that does not have proxy time can be bad, because we can skip a fault in the bookinfo installation. So We decided to make the validation in each pod with proxy
2. It's easier to read to anyone when you make the validation in this way

File: pkg/util/oc/oc_struct.go (hunk 24 above)
Comment (fjglira) [reply pkg/util/oc/oc_struct.go:432]: I'll change this

**Comment by fjglira** [comment]
Updated, I disable the delete smcp subtest and open a bug related to the route that it's not being deleted to verify this case

File: pkg/util/oc/oc_struct.go (hunk 24 above)
Comment (fjglira) [reply pkg/util/oc/oc_struct.go]: I added a scape for single quotes

File: pkg/tests/ossm/smoke_test.go [hunk 27]
```diff
@@ -79,3 +173,39 @@ func assertSidecarInjectedInAllBookinfoPods(t TestHelper, ns string) {
...
+// 	oc.DeleteFromString(t, meshNamespace, GetSMMRTemplate())
+// 	DeleteSMCPVersion(t, meshNamespace, ver)
+// 	retry.UntilSuccess(t, func(t TestHelper) {
+// 		oc.GetAllResources(t,
+// 			meshNamespace,
+// 			assert.OutputContains("No resources found in",
+// 				"All resources deleted from namespace",
+// 				"Still waiting for resources to be deleted from namespace"))
+// 	})
+// }
```
Comment (jewertow) [inline pkg/tests/ossm/smoke_test.go:197]: Why not to remove this function? Will you remember about it?
```suggestion
```

**Review by jewertow — APPROVED** [review]

File: pkg/tests/ossm/smoke_test.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (jewertow) [reply pkg/tests/ossm/smoke_test.go]: @luksa I suggested to simplify it, because it looked more complicated and I had to execute it to understand what it does in fact. It also required more complicated parsing. I didn't know that you suggested to use this command.

Anyway, now the code is simpler to read and understand.

File: pkg/tests/ossm/smoke_test.go (hunk 27 above)
Comment (fjglira) [reply pkg/tests/ossm/smoke_test.go:197]: Hi @jewertow I did not remove because is used in the sub test case that I disable because of this bug: https://issues.redhat.com/browse/OSSM-4064

File: pkg/tests/ossm/smoke_test.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (luksa) [reply pkg/tests/ossm/smoke_test.go:151]: I agree. The reason I asked was because I remember @fjglira increasing this from 5s to 10s, because 5s wasn't enough.
